src/plugins/plugin-http.js

```diff
@@ -0,0 +1,124 @@
+  // In theory we should use url.format here. However, that is
+  // broken. See: https://github.com/joyent/node/issues/9117 and
+  // https://github.com/nodejs/io.js/pull/893
+  // Let's do things the same way _http_client does it.
+  return isString(uri) ? uri :
+    (options.protocol || agent.protocol) + '//' +
+    (options.hostname || options.host || 'localhost') +
+    ((isString(options.port) ? (':' + options.port) : '')) +
+    (options.path || options.pathName || '/');
+}
+
+function patchRequest (http, api) {
+  return shimmer.wrap(http, 'request', function requestWrap(request) {
+    return function request_trace(options, callback) {
+      if (!options) {
+        return request.apply(this, arguments);
+      }
+
+      // Don't trace ourselves lest we get into infinite loops
+      // Note: this would not be a problem if we guarantee buffering
+      // of trace api calls. If there is no buffering then each trace is
+      // an http call which will get a trace which will be an http call
+      if (isTraceAgentRequest(options)) {
+        return request.apply(this, arguments);
+      }
+
+      options = isString(options) ? url.parse(options) : merge({}, options);
+      options.headers = options.headers || {};
+
+      var uri = extractUrl(options);
+      var requestLifecycleSpan =
+          api.createChildSpan({name: getSpanName(options)});
+      if (!requestLifecycleSpan) {
+        return request.apply(this, arguments);
+      }
+
+      requestLifecycleSpan.addLabel(api.labels.HTTP_METHOD_LABEL_KEY,
+                                    options.method);
+      requestLifecycleSpan.addLabel(api.labels.HTTP_URL_LABEL_KEY, uri);
+      options.headers[api.constants.TRACE_CONTEXT_HEADER_NAME] =
+          requestLifecycleSpan.getTraceContext();
+      var req = request.call(this, options, function(res) {
+        api.wrapEmitter(res);
+        var numBytes = 0;
+        res.on('data', function (chunk) {
+          numBytes += chunk.length;
+        });
+        res.on('end', function () {
+          requestLifecycleSpan
+            .addLabel(api.labels.HTTP_RESPONSE_SIZE_LABEL_KEY, numBytes);
+          requestLifecycleSpan
+            .addLabel(api.labels.HTTP_RESPONSE_CODE_LABEL_KEY, res.statusCode);
+          requestLifecycleSpan.endSpan();
+        });
+        if (callback) {
+          return callback(res);
+        }
+      });
+      api.wrapEmitter(req);
+      req.on('error', function (e) {
+        if (e) {
+          requestLifecycleSpan.addLabel(api.labels.ERROR_DETAILS_NAME, e.name);
+          requestLifecycleSpan
+            .addLabel(api.labels.ERROR_DETAILS_MESSAGE, e.message);
+        } else {
+          // What's the new logger target?
+          // console.error('HTTP request error was null or undefined', e);
+        }
+        requestLifecycleSpan.endSpan();
+      });
+      return req;
+    };
+  });
+
+  function isTraceAgentRequest (options) {
+    return options && options.headers &&
```

Comment:
This will only work if `options` is an object and not a string. I think we're fine because all of the requests made by our agent use an object but it may be worth generalizing or parsing string options in a follow on (especially since the old http hook had this same logic).